### PR TITLE
feat: gadgetize `VerifyProof`

### PR DESCRIPTION
**Pull request by Eagle941**
This PR integrates the library `reilabs/gnark-lean-extractor` to formally verify in Lean4 the correctness of circuit `VerifyProof`.

A new CLI flag `extract-circuit` is introduced with parameter `tree-depth` to autogenerate the Lean code.

### CODE REVIEW

File: formal-verification/FormalVerification.lean [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import ProvenZk.Gates
 import ProvenZk.Ext.Vector
 
-namespace Poseidon2
+namespace VerifyProof
```
Comment (kustosz) [inline formal-verification/FormalVerification.lean:4]: This changing is annoying (will happen every time you swap gadgets) – we should allow injecting this namespace name when exporting the circuit and set it manually in this project, given the incremental nature of this.

File: main.go [hunk 2]
```diff
@@ -239,11 +239,13 @@ func main() {
 				Name: "extract-circuit",
 				Flags: []cli.Flag{
 					&cli.StringFlag{Name: "output", Usage: "Output file", Required: true},
+					&cli.StringFlag{Name: "proof-size", Usage: "Length of Proof vector (i.e. depth of tree)", Required: true},
```
Comment (kustosz) [inline main.go:242]: This should follow the naming convention of the rest of this and be called `tree-depth`

File: prover/circuit.go [hunk 3]
```diff
@@ -160,12 +176,16 @@ func (circuit *MbuCircuit) Define(api frontend.API) error {
 		currentIndex := api.Add(circuit.StartIndex, i)
 		currentPath := api.ToBinary(currentIndex, circuit.Depth)
 
+		// len(circuit.MerkleProofs[i]) === circuit.Depth
```
Comment (kustosz) [inline prover/circuit.go:179]: commented out code

**Review by kustosz — CHANGES_REQUESTED** [review]

File: prover/proving_system.go [hunk 4]
```diff
@@ -115,8 +114,9 @@ func Setup(treeDepth uint32, batchSize uint32) (*ProvingSystem, error) {
 	return &ProvingSystem{treeDepth, batchSize, pk, vk, ccs}, nil
 }
 
-func ExtractLean() (string, error) {
-	assignment := poseidon.Poseidon2{}
+func ExtractLean(proofSize uint32) (string, error) {
+	// Not checking for proofSize === 0
```
[the remark this answers is not in the document]
Comment (Eagle941) [reply prover/proving_system.go:118]: Should it be possible to write a circuit with treeDepth of 0?

File: prover/circuit.go (hunk 3 above)
Comment (Eagle941) [reply prover/circuit.go:179]: I added these as comments for reference. Should I delete?

**Review by kustosz — APPROVED** [review]

**Pull request merged**